File: 02-week-2--recursion/04-day-4--recursive-find-shortest-string/solutions/benchmark/benchmark.rb
def find_shortest_string(arr)
  arr.reduce do |shortest, string|
    string.length < shortest.length ? string : shortest
  end
end

def find_shortest_string_recursive(arr)
  return arr.first if arr.length == 1

  result = find_shortest_string_recursive(arr[1..-1])

  arr.first.length <= result.length ? arr.first : result
end

def benchmark
  start_time = Time.now

  1000.times do 
    yield
  end

  (Time.now - start_time) / 1000
end

puts "Iterative: #{benchmark { find_shortest_string(['cat', 'dogs', '', 'bats', 'flags']) }}"
puts "Recursive: #{benchmark { find_shortest_string_recursive(['cat', 'dogs', '', 'bats', 'flags']) }}"

File: 02-week-2--recursion/04-day-4--recursive-find-shortest-string/solutions/benchmark/benchmark.js
function findShortestString(arr) {
  return arr.reduce((shortest, string) =>
    string.length < shortest.length ? string : shortest
  );
}

function findShortestStringRecursive(arr) {
  if (arr.length === 1) {
    return arr[0];
  }

  const result = findShortestStringRecursive(arr.slice(1));

  return arr[0].length <= result.length ? arr[0] : result;
}

function benchmark(callback) {
  const startTime = Date.now();

  for (let i = 0; i < 1000; ++i) {
    callback();
  }

  return (Date.now() - startTime) / 1000;
}

console.log('Iterative:', benchmark(() => findShortestString(['cat', 'dogs', '', 'bats', 'flags'])));
console.log('Recursive:', benchmark(() => findShortestStringRecursive(['cat', 'dogs', '', 'bats', 'flags'])));